### ISSUE TITLE
Callee still doesnot properly validate if the action exists

### ISSUE THREAD

**Issue by sanzeeb3** [woocommerce/action-scheduler#580]
Referring to this statement:

https://github.com/woocommerce/action-scheduler/blob/c98e384c6401edde7d57d97b6d3681e34353c0db/classes/ActionScheduler_ListTable.php#L471

Failing to validate the action is resulting in the error if the canceled action is attempted to delete manually.  It seems the canceled action is deleted on the background, so if it is attempted to delete, the error appears because the action doesn't exist anymore.

To reproduce the error, cancel an action, wait for a minute, attempt to delete bulk actions, you'll get an error:

> [17-Jun-2020 05:17:23 UTC] PHP Fatal error:  Uncaught InvalidArgumentException: Unidentified action 2 in C:\laragon\www\myenv\wp-content\plugins\action-scheduler\classes\data-stores\ActionScheduler_DBStore.php:548

**Comment by rrennick** [comment]
> To reproduce the error, cancel an action, wait for a minute, attempt to delete bulk actions, you'll get an error:

@sanzeeb3 These steps don't produce an error in my install. Can you post your WooCommerce System Status Report?

**Comment by sanzeeb3** [comment]
@rrennick, actually I don't have WooCommerce installed. but that error is obvious when you attempt to delete the action that does not exist anymore. So, to reproduce: stay on 'All', cancel a pending action, the same action appears with 'cancelled' status, wait for a minute (I don't know exactly when cancelled action gets automatically deleted on the background), delete that (or all) actions.

**Comment by rrennick** [comment]
> I don't know exactly when cancelled action gets automatically deleted on the background

Canceled actions are purged 30 days after their schedule date. Do you have a filter set up for them to be removed right away?

**Comment by sanzeeb3** [comment]
It seems that it isn't happening. Canceled actions are being removed after 30 seconds or so (or even immediately). I don't have any filter setup for that. I also checked with only AC active where I canceled a pending `action_scheduler/migration_hook` action and delete after a few seconds. Even with that, it might be logical to check if action exists because, after 30 days when you are visiting the page and the action is being deleted in the background at that very time, the error will appear when you attempt to delete it manually.

**Comment by rrennick** [comment]
Action Scheduler does remove the `action_scheduler/migration_hook` action when the migration is complete.

**Comment by sanzeeb3** [comment]
I mean the issue is for all canceled actions, not completed. `action_scheduler/migration_hook` is only for an example. if you cancel this and attempt to delete after some time, you'll see a fatal error.

**Comment by rrennick** [comment]
Can you check your `wp_options` table for a `action_scheduler_migration_status` setting?

**Comment by sanzeeb3** [comment]
yeah, `action_scheduler_migration_status` is complete. but that doesn't justify anything. I mean this issue for every canceled status. You can reproduce the issue on every install. I repeat the steps. Stay in 'All', 'Cancel a pending action', 'wait for a few time', 'delete that(or all)' action. I've not used any filter.

> Canceled actions are purged 30 days after their scheduled date. 

This isn't happening. Canceled actions are removed after a few seconds (not days). Now that when you attempt to delete the action which is already removed, the error above is obvious.

**Comment by rrennick** [comment]
@sanzeeb3 I tested with the following

- create a new test action pending for tomorrow
- went to the last page of All actions
- canceled the action
- waited approximately 15 minutes
- checked just that canceled action
- deleted through the bulk actions

No errors on screen or in the error log and the action was deleted.

**Comment by sanzeeb3** [comment]
@rrennick oh. For context, I'm creating a recurring action if that is any exception, with the following.

```
as_schedule_recurring_action( strtotime( '+3 day' ), DAY_IN_SECONDS, 'schedule_emails', array( 'example@example.com', 'subject', 'message' ), 'schedule_emails' );
```
Please let me know if there's anything wrong with that. I'll debug the whole issue later.

**Comment by rrennick** [comment]
@sanzeeb3 I tried with a recurring action and had the same result. Have you tried this in an install with only Action Scheduler active?

**Comment by sanzeeb3** [comment]
yes, I tried with a fresh WP install with only Action Scheduler, and that code snippet active. Actually, I see this issue on every other install as well. I'll get details about the issue later. Thanks!

**Comment by sanzeeb3** [comment]
@rrennick 

> Canceled actions are purged 30 days after their scheduled date.

It seems canceled and completed actions are purged 30 days after the modified date, and not scheduled date [see here](https://github.com/woocommerce/action-scheduler/blob/master/classes/ActionScheduler_QueueCleaner.php#L44). The cutoff date (30 days) is compared with the last_attempt_gmt. The completed actions have the last_attempt_gmt, but canceled actions don't (https://a.supportally.com/YnVX9e). With this, the cutoff date is always greater than last_attempt_gmt for canceled actions and are therefore deleted on the background. So, either the cutoff date should be compared with scheduled date (rather than modified date) or the canceled actions should have last_attempt_gmt. In either way, this issue is still valid, the existence of the action should be checked before deletion. Lastly, I'm not sure why you can't reproduce the issue on the clean install.

**Comment by rrennick** [comment]
> The cutoff date (30 days) is compared with the last_attempt_gmt. 

@sanzeeb3 Thank you for tracking that down.

**Comment by rrennick** [comment]
Addressed in #582
